tutorials/docs-17-implementing-samplers/index.qmd

```diff
@@ -0,0 +1,479 @@
+
+
+## Quick overview of MALA
+
+We can view MALA as a single step of the leapfrog intergrator with resampling of momentum $p$ at every step. To make that statement a bit more concrete, we first define the *extended* target $\bar{\gamma}(x, p)$ as
+
+\begin{equation*}
+\log \bar{\gamma}(x, p) \propto \log \gamma(x) + \log \gamma_{\mathcal{N}(0, M)}(p)
+\end{equation*}
+
+where $\gamma_{\mathcal{N}(0, M)}$ denotes the density for a zero-centered Gaussian with covariance matrix $M$.
+We then consider targeting this joint distribution over both $x$ and $p$ as follows.
+First we define the map
+
+\begin{equation*}
+\begin{split}
+  L_{\epsilon}: \quad & \mathbb{R}^d \times \mathbb{R}^d \to \mathbb{R}^d \times \mathbb{R}^d \\
+  & (x, p) \mapsto (\tilde{x}, \tilde{p}) := L_{\epsilon}(x, p)
+\end{split}
+\end{equation*}
+
+as
+
+\begin{equation*}
+\begin{split}
+  p_{1 / 2} &:= p + \frac{\epsilon}{2} \nabla \log \gamma(x) \\
+  \tilde{x} &:= x + \epsilon M^{-1} p_{1 /2 } \\
+  p_1 &:= p_{1 / 2} + \frac{\epsilon}{2} \nabla \log \gamma(\tilde{x}) \\
+  \tilde{p} &:= - p_1
+\end{split}
+\end{equation*}
+
+This might be familiar for some readers as a single step of the Leapfrog integrator.
+We then define the MALA kernel as follows: given the current iterate $x_i$, we sample the next iterate $x_{i + 1}$ as
+
+\begin{equation*}
+\begin{split}
+  p &\sim \mathcal{N}(0, M) \\
+  (\tilde{x}, \tilde{p}) &:= L_{\epsilon}(x_i, p) \\
+  \alpha &:= \min \left\{ 1, \frac{\bar{\gamma}(\tilde{x}, \tilde{p})}{\bar{\gamma}(x_i, p)} \right\} \\
+  x_{i + 1} &:=
+  \begin{cases}
+    \tilde{x} \quad & \text{ with prob. } \alpha \\
+    x_i       \quad & \text{ with prob. } 1 - \alpha
+  \end{cases}
+\end{split}
+\end{equation*}
+
+i.e. we accept the proposal $\tilde{x}$ with probability $\alpha$ and reject it, thus sticking with our current iterate, with probability $1 - \alpha$.
+
+## What we need from a model: [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl)
+
+There are a few things we need from the "target" / "model" / density that we want to sample from:
+
+1.  We need access to log-density *evaluations* $\log \gamma(x)$ so we can compute the acceptance ratio involving $\log \bar{\gamma}(x, p)$.
+2.  We need access to log-density *gradients* $\nabla \log \gamma(x)$ so we can compute the Leapfrog steps $L_{\epsilon}(x, p)$.
+3.  We also need access to the "size" of the model so we can determine the size of $M$.
+
+Luckily for us, there is a package called [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) which provides an interface for *exactly* this!
+
+To demonstrate how one can implement the "[LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) interface"[^1] we will use a simple Gaussian model as an example:
+
+```{julia}
+using LogDensityProblems: LogDensityProblems;
+
+# Let's define some type that represents the model.
+struct NormalLogDensity{M<:AbstractVector{<:Real}}
```

Comment:
Might it make sense to call this something else, like `IsotropicNormalModel` or something? It took me a moment to understand that this represents just the Gaussian model for which we can compute log densities, not a log density calculation.

Reply:
Good idea:)